adam-core/src/main/scala/org/bdgenomics/adam/util/MdTag.scala

```diff
@@ -42,72 +42,105 @@ object MdTag {
+            }
+            // Move the reference position the length of the matching sequence
+            referencePos += numMatchingBases.toInt
+            if (matchingBases.toInt <= cigarElement.getLength) mdTagStringOffset += matchingBases.size
+
+            // If the M operator has been fully read move on to the next operator
+            if (referencePos >= cigarReferencePosition + cigarElement.getLength) {
+              cigarIdx += 1
+              cigarReferencePosition += cigarElement.getLength
             }
-            offset += s.length
-            referencePos += length
-        }
-      }
-
-      readMatches("MD tag must start with a digit")
-
-      while (offset < end) {
-        val mdTagType = {
-          if (mdTag.charAt(offset) == '^') {
-            offset += 1
-            MdTagEvent.Delete
-          } else {
-            MdTagEvent.Mismatch
           }
-        }
-        basesPattern.findPrefixOf(mdTag.substring(offset)) match {
-          case None => throw new IllegalArgumentException("Failed to find deleted or mismatched bases after a match: %s".format(mdTagInput))
-          case Some(bases) =>
-            mdTagType match {
-              case MdTagEvent.Delete =>
-                bases.foreach {
-                  base =>
-                    deletions += (referencePos -> base)
-                    referencePos += 1
-                }
-              case MdTagEvent.Mismatch =>
-                bases.foreach {
+          case (CigarOperator.MATCH_OR_MISMATCH | CigarOperator.M | CigarOperator.X, None) => {
+            basesPattern.findPrefixOf(mdTag.substring(mdTagStringOffset)) match {
+              // Must have found a base or digit pattern in CigarOperator M
+              case None => throw new IllegalArgumentException(s"No match or mismatched bases found for ${cigar.toString} in MDTag $mdTag")
+              case Some(mismatchedBases) => {
+                mismatchedBases.foreach {
                   base =>
                     mismatches += (referencePos -> base)
                     referencePos += 1
                 }
+                mdTagStringOffset += mismatchedBases.size
+              }
+            }
+            // If the M operator has been fully read move on to the next operator
+            if (referencePos >= cigarReferencePosition + cigarElement.getLength) {
+              cigarIdx += 1
+              cigarReferencePosition += cigarElement.getLength
+            }
+          }
+          case (CigarOperator.DELETION, None) => {
+            mdTag.charAt(mdTagStringOffset) match {
+              case '^' => {
+                // Skip ahead of the deletion '^' character
+                mdTagStringOffset += 1
+                basesPattern.findPrefixOf(mdTag.substring(mdTagStringOffset)) match {
+                  case None => throw new IllegalArgumentException(s"No deleted bases found ${cigar.toString} in MDTag $mdTag")
+                  case Some(deletedBases) => {
```

Comment:
sanity check: should `deletedBases.length == cigarElement.getLength` here? i.e. the number of bases after the `^` in the MDTag should match the number prefixing the cigar operator `D`?